### PR TITLE
Use stubfile extension for PasswordUpgraderInterface

### LINKED ISSUE #333: Generics for PasswordUpgraderInterface fail with SF < 5.3
Hi @VincentLanglet, the commit https://github.com/phpstan/phpstan-symfony/commit/dac2474904b092267f0a19dfba8c46b6f21eab6a caused a BC, which I  can't solve. I would appreciate your input 😄 

A Symfony project that still uses 4.4 fails now with 
```
Parameter $user of method App\Security\DoctrineUserProvider::upgradePassword() has invalid type Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface. 
```

This is the Symfony 4.4 code:
https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Security/Core/User/PasswordUpgraderInterface.php

The interface that you check was first added with 5.3, see the sources at 5.2:
https://github.com/symfony/symfony/tree/5.2/src/Symfony/Component/Security/Core/User

Is that behavior somehow configurable (e.g. add a SF version in phpstan.neon config) or do I have to pin my composer dep?
```
        "phpstan/phpstan-symfony": "1.2.18",
```

In my case it doesn't matter, as this project branch only receives security fixes ... but considering that the `PasswordAuthenticatedUserInterface` interface was first added with 5.3 this might be an issue for others as well.

I don't know how the stubs work, but could it help if your stubbed interface would extend `UserInterface` to make it compatible with old SF versions?

### PR DESCRIPTION
**Pull request by VincentLanglet**
I'm trying to solve https://github.com/phpstan/phpstan-symfony/issues/333
@kevinpapst can you try this branch ?

Closes https://github.com/phpstan/phpstan-symfony/issues/333

### CODE REVIEW

**Comment by kevinpapst** [comment]
@VincentLanglet Thank you! Works for SF 4.4. But with SF 6.2 I have other issues now.

Below output is from the SF 4.4 project, Symfony 6.2 follows later:

Below:
```bash
$ vendor/bin/phpstan analyse src -c phpstan.neon --level=5
 951/951 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Security/DoctrineUserProvider.php                                                                                                                                         
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  80     Parameter $user of method App\Security\DoctrineUserProvider::upgradePassword() has invalid type Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface.  
 ------ -------------------------------------------------------------------------------------------------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  Line   Security/KimaiUserProvider.php                                                                                                                                         
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
  87     Parameter $user of method App\Security\KimaiUserProvider::upgradePassword() has invalid type Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface.  
 ------ ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- 
                                                                                                                        
 [ERROR] Found 2 errors                                                                                                 

Script vendor/bin/phpstan analyse src -c phpstan.neon --level=5 handling the phpstan event returned with error code 1
```

Composer update:
```bash
$ composer update 
Loading composer repositories with package information                                                                                                                                                     Restricting packages listed in "symfony/symfony" to "4.4.*"
Updating dependencies
Lock file operations: 0 installs, 1 update, 0 removals
  - Upgrading phpstan/phpstan-symfony (1.2.23 => dev-PasswordAuthenticatedUserStubFilesExtension d879c1a)
Writing lock file
Installing dependencies from lock file (including require-dev)
Package operations: 0 installs, 1 update, 0 removals
  - Downloading phpstan/phpstan-symfony (dev-PasswordAuthenticatedUserStubFilesExtension d879c1a)
  - Upgrading phpstan/phpstan-symfony (1.2.23 => dev-PasswordAuthenticatedUserStubFilesExtension d879c1a): Extracting archive
Package sensio/framework-extra-bundle is abandoned, you should avoid using it. Use Symfony instead.
Package symfony/debug is abandoned, you should avoid using it. Use symfony/error-handler instead.
Package symfony/inflector is abandoned, you should avoid using it. Use EnglishInflector from the String component instead.
Package php-cs-fixer/diff is abandoned, you should avoid using it. No replacement was suggested.
Generating autoload files
composer/package-versions-deprecated: Generating version class...
composer/package-versions-deprecated: ...done generating version class
139 packages you are using are looking for funding.
Use the `composer fund` command to find out more!
```

Afterwards:
```bash
$ vendor/bin/phpstan analyse src -c phpstan.neon --level=5
 951/951 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%
                                                                                                                        
 [OK] No errors                                                                                                         

```

**Comment by VincentLanglet** [comment]
> @VincentLanglet Thank you! Works for SF 4.4.

Great

> But with SF 6.2 I have other issues now.

Certainly because I used `class_exists` on an interface so the stub wasn't loaded anymore.
I added a commit to fix it.

**Comment by kevinpapst** [comment]
Yep 👍  The lastest commit fixed the SF 6.2 issues. And SF 4.4 still works.

I assume your branch is behind 1.2.23, right? 

Because I get `PHPDoc tag @extends contains generic type Symfony\Component\Security\Core\Authorization\Voter\Voter<string, App\Entity\Activity> but class Symfony\Component\Security\Core\Authorization\Voter\Voter is not generic.` in SF 6.2, which worked in 1.2.23.

**Comment by VincentLanglet** [comment]
> Yep 👍 The lastest commit fixed the SF 6.2 issues. And SF 4.4 still works.

Great !

> I assume your branch is behind 1.2.23, right?
> 
> Because I get `PHPDoc tag @extends contains generic type Symfony\Component\Security\Core\Authorization\Voter\Voter<string, App\Entity\Activity> but class Symfony\Component\Security\Core\Authorization\Voter\Voter is not generic.` in SF 6.2, which worked in 1.2.23.

Certainly. I rebased my branch ;)

**Review by kevinpapst — APPROVED** [review]
Alright, now SF 6.2 is green as well. Fixed 👍 

Thank you for your quick support, very much appreciated!

File: src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+use PHPStan\PhpDoc\StubFilesExtension;
+use function interface_exists;
+
+class PasswordAuthenticatedUserStubFilesExtension implements StubFilesExtension
+{
+
+	public function getFiles(): array
+	{
+		if (!interface_exists('Symfony\Component\Security\Core\User\PasswordAuthenticatedUserInterface')) {
```
Comment (ondrejmirtes) [inline src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php]: Runtime concerns are irrelevant for PHPStan, you need to ask ReflectionProvider. If ReflectionProvider cannot be injected, then ask Reflector, as phpstan-doctrine does: https://github.com/phpstan/phpstan-doctrine/blob/1.3.x/src/Stubs/Doctrine/StubFilesExtensionLoader.php

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php (hunk 1 above)
Comment (kevinpapst) [reply src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php]: Not that I understand anything about PHPStan internals, but isn't that the "same" as in : https://github.com/phpstan/phpstan-symfony/blob/1.2.x/src/Symfony/InputBagStubFilesExtension.php#L13 ?

Comment (ondrejmirtes) [reply src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php]: That is reaaaaaally old and should be fixed too.

Comment (VincentLanglet) [reply src/Symfony/PasswordAuthenticatedUserStubFilesExtension.php]: Done for both StubFilesExtension.

**Comment by VincentLanglet** [comment]
Friendly ping @ondrejmirtes, is this PR ok to you ?
Thanks.

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**